File: pagerank.c
#include <stdio.h>
#include <math.h>
#include <string.h>
#include <stdlib.h>

typedef struct{
	int nodeCount;
	int **outNodes;
	int *outDegree;
	double *currentPageRankScore;
	double *newPageRankScore;
	double dampingFactor;
	double epision;
} Graph;

Graph* newGraph(int nodeCount, double dampingFactor, double epision){
	Graph *graph = (Graph*)malloc(sizeof(Graph));

	graph->dampingFactor = dampingFactor;
	graph->epision = epision;

	graph->nodeCount = nodeCount;
	graph->outNodes = (int**)malloc(sizeof(int*) * (nodeCount + 1));
	graph->outDegree = (int*)malloc(sizeof(int) * (nodeCount + 1));
	graph->currentPageRankScore = (double*)malloc(sizeof(double) * (nodeCount + 1));
	graph->newPageRankScore = (double*)malloc(sizeof(double) * (nodeCount + 1));

	for(int v = 1; v <= nodeCount; v ++){
		graph->outNodes[v] = NULL;
		graph->outDegree[v] = 0;
	}

	return graph;
}

void deleteGraph(Graph **graphPointer){
	if(graphPointer == NULL || *graphPointer == NULL){
		return;
	}
	Graph *graph = *graphPointer;
	free(graph->outDegree);
	free(graph->currentPageRankScore);
	free(graph->newPageRankScore);
	
	for(int v = 0; v <= graph->nodeCount; v ++){
		if(graph->outNodes[v] != NULL){
			free(graph->outNodes[v]);
		}
	}

	free(graph->outNodes);

	free(graph);
	*graphPointer = NULL;
}

int scoreNoDifference(double *score1, double *score2, int dimension, double threshold){
	double distance = 0;

	for(int i = 1; i <= dimension; i ++){
		double difference = score1[i] - score2[i];
		distance += difference * difference;
	}
	
	distance = sqrt(distance);
	printf("\t\tDistance %f\n", distance);
	return (distance < threshold) ? 1 : 0;
}

// Runs ths PageRank algorithm on the graph.
// The converged scores of nodes is stored in the variable currentPageRankScore.
void graphRunPageRank(Graph *graph){
	// Initialize the scores.
	for(int v = 1; v <= graph->nodeCount; v ++){
		graph->currentPageRankScore[v] = 1;
		graph->newPageRankScore[v] = 0;
	}

	for(int i = 1; ; i ++){
		printf("\tIteration %d\n", i);

		// Vote the scores to the neighbors of the nodes.
		// Pre-compute the overall score of the zero-out-degree nodes.
		double zeroDegreeScore = 0;

		for(int v = 1; v <= graph->nodeCount; v ++){
			if(graph->outDegree[v] == 0){
				zeroDegreeScore += graph->currentPageRankScore[v];
			}
			else{
				double voteScore = graph->currentPageRankScore[v] / graph->outDegree[v];

				for(int x = 0; x < graph->outDegree[v]; x ++){
					int u = graph->outNodes[v][x];

					graph->newPageRankScore[u] += voteScore;
				}
			}
		}

		zeroDegreeScore /= graph->nodeCount;
		
		// Adjust the scores with the damping factor.
		for(int v = 1; v <= graph->nodeCount; v ++){
			graph->newPageRankScore[v] = (1.0 - graph->dampingFactor) + graph->dampingFactor * (zeroDegreeScore + graph->newPageRankScore[v]);
		}

		// Break the iterations if the scores are converged.
		if(scoreNoDifference(graph->newPageRankScore, graph->currentPageRankScore, graph->nodeCount, graph->epision)){
			break;
		}

		// Update the scores.
		for(int v = 1; v <= graph->nodeCount; v ++){
			graph->currentPageRankScore[v] = graph->newPageRankScore[v];
			graph->newPageRankScore[v] = 0;
		}
	}
}

int main(int argc, char *argv[]){
	double dampingFactor = 0.0;
	double epision = 0.0;
	char inputFileName[100];
	char outputFileName[100];
	
	for(int a = 1; a < argc; a ++){
		if(argv[a][0] == '-'){
			switch(argv[a][1]){
				case 'd':
					a ++;
					sscanf(argv[a], "%lf", &dampingFactor);
					break;
				case 'e':
					a ++;
					sscanf(argv[a], "%lf", &epision);
					break;
				case 'o':
					a ++;
					strcpy(outputFileName, argv[a]);
					break;
			}
		}
		else{
			strcpy(inputFileName, argv[a]);
			break;
		}
	}

	printf("output File %s, input File %s\n", outputFileName, inputFileName);

	printf("Read data\n");
	FILE *inputFile = fopen(inputFileName, "r");
	
	char maxNode[100];
	int nodeCount;
	fscanf(inputFile, "%s%d", maxNode, &nodeCount);
	Graph *graph = newGraph(nodeCount, dampingFactor, epision);
	
	printf("Node count %d, damping factor %f, epision %f\n", nodeCount, dampingFactor, epision);

	int node;
	int outDegree;
	while(fscanf(inputFile, "%d:%d", &node, &outDegree) == 2){
		graph->outDegree[node] = outDegree;
		graph->outNodes[node] = (int*)malloc(sizeof(int) * outDegree);

		for(int u = 0; u < outDegree; u ++){
			fscanf(inputFile, "%d", &graph->outNodes[node][u]);
		}
	}
	
	fclose(inputFile);

	int zeroDegree = 0;
	for(int v = 1; v <= graph->nodeCount; v ++){
		if(graph->outDegree[v] == 0){
			zeroDegree ++;
		}
	}
	printf("Zero-out-degree nodes: %d\n", zeroDegree);

	printf("Run PageRank\n");
	graphRunPageRank(graph);

	printf("Write results\n");
	FILE *outputFile = fopen(outputFileName, "w");
	
	for(int v = 1; v <= graph->nodeCount; v ++){
		fprintf(outputFile, "%d:%f\n", v, graph->currentPageRankScore[v]);
	}

	fclose(outputFile);

	deleteGraph(&graph);
	
	return 0;
}

File: pagerank.py
import sys;
import math;

class Graph:
    def __init__(self, nodeCount, dampingFactor, epision):
        self.nodeCount = nodeCount;
        self.dampingFactor = dampingFactor;
        self.epision = epision;
        self.nodeDict = {i: {"outNodes": [], "degree": 0, "currentPageRankScore": 1.0, "newPageRankScore: 0.0"} for i in xrange(1, nodeCount + 1)};

    def addArc(self, nodeID, outNodeID):
        self.nodeDict[nodeID]["outNodes"].append(outNodeID);
        self.nodeDict[nodeID]["degree"] += 1;

    def needHalt(self, allNodes):
        distance = 0.0;
        
        for node in allNodes:
            difference = node.getPageRankScoreDifference();
            distance += difference * difference;

        distance = math.sqrt(distance);

        print "\t\tDistance ", distance;

        return distance < self.epision;

    def runPageRank(self):
        allNodes = self.nodeDict.items();
        
        for (id, node) in allNodes:
            node["currentPageRankScore"] = 1.0;
            node["newPageRankScore"] = 0.0;

        iteration = 1;

        while 1:
            print "\tIteration ", iteration;
    
            print "\tVote scores";
            for (id, node) in allNodes:
                if id % 1000 == 1:
                    print "\t\tNode ", id;

                if node["degree"] == 0:
                    voteScore = node["currentPageRankScore"] / len(allNodes);
                    for (outID, outNode) in allNodes:
                        outNode["newPageRankScore"] += voteScore;
                else:

            print "\tDamp scores";
            for node in allNodes:
                node.dampNewPageRankScore(self.dampingFactor);

            print "\tJudge threshold";
            if self.needHalt(allNodes):
                break;

            print "\tUpdate scores";
            for node in allNodes:
                node.updatePageRankScore();

            iteration += 1;

    def writePageRankScores(self, outputFile):
        for (id, node) in sorted(self.nodeDict.items()):
            outputFile.write(str(id) + ":" + str(node.getPageRankScore()) + "\n");

def main():
    dampingFactor = 0.0;
    epision = 0.0;
    outputName = "";
    inputFileName = "";

    a = 1;

    while 1:
        if a >= len(sys.argv) + 1:
            break;

        if sys.argv[a] == "-d":
            a += 1;
            dampingFactor = float(sys.argv[a]);
        elif sys.argv[a] == "-e":
            a += 1;
            epision = float(sys.argv[a]);
        elif sys.argv[a] == "-o":
            a += 1;
            outputFileName = sys.argv[a];
        else:
            inputFileName = sys.argv[a];
            break;

        a += 1;

    print "d: ", dampingFactor;
    print "e: ", epision;

    print "Read the file and construct the graph";

    inputFile = open(inputFileName, "r");
    nodeCount = int(inputFile.readline().split()[1]);
    graph = Graph(nodeCount, dampingFactor, epision);

    print "Number of nodes: ", nodeCount;

    while 1:
        line = inputFile.readline();
        if not line:
            break;

        outNodeIDs = line.split();
        firstOutNodeIDPair = outNodeIDs[0].split(":");
        nodeID = int(firstOutNodeIDPair[0]);
        firstOutNodeID = int(firstOutNodeIDPair[1]);
        graph.addArc(nodeID, firstOutNodeID);

        for (i, outNodeIDString) in enumerate(outNodeIDs):
            if i == 0:
                continue;

            outNodeID = int(outNodeIDString); 
            graph.addArc(nodeID, outNodeID);
        
    inputFile.close();

    print "Start PageRank";

    graph.runPageRank();

    print "Write results";

    outputFile = open(outputFileName, "w");
    graph.writePageRankScore(outputFile);
    outputFile.close();
    
if __name__ == "__main__":
    main();
